src/framework/scene-grab.js

```diff
@@ -0,0 +1,363 @@
+import {
+    ADDRESS_CLAMP_TO_EDGE,
+    CLEARFLAG_COLOR, CLEARFLAG_DEPTH,
+    FILTER_NEAREST, FILTER_LINEAR, FILTER_LINEAR_MIPMAP_LINEAR,
+    PIXELFORMAT_DEPTHSTENCIL, PIXELFORMAT_R8_G8_B8_A8, PIXELFORMAT_R8_G8_B8
+} from '../graphics/constants.js';
+
+import { RenderTarget } from '../graphics/render-target.js';
+import { Texture } from '../graphics/texture.js';
+import { DebugGraphics } from '../graphics/debug-graphics.js';
+
+import {
+    LAYERID_DEPTH, LAYERID_WORLD,
+    SHADER_DEPTH
+} from '../scene/constants.js';
+
+import { Layer } from '../scene/layer.js';
+
+/** @typedef {import('../graphics/graphics-device.js').GraphicsDevice} GraphicsDevice */
+/** @typedef {import('./components/camera/component.js').CameraComponent} CameraComponent */
+
+// Internal class abstracting the access to the depth and color texture of the scene.
+// color frame buffer is copied to a texture
+// For webgl 2 devices, the depth buffer is copied to a texture
+// for webgl 1 devices, the scene's depth is rendered to a separate RGBA texture
+//
+// TODO: implement mipmapped color buffer support for WebGL 1 as well, which requires
+// the texture to be a power of two, by first downscaling the captured framebuffer
+// texture to smaller power of 2 texture, and then generate mipmaps and use it for rendering
+// TODO: or even better, implement blur filter to have smoother lower levels
+class SceneGrab {
+    constructor(application) {
+        this.application = application;
+
+        /** @type {GraphicsDevice} */
+        this.device = application.graphicsDevice;
+
+        // create depth layer
+        this.clearOptions = null;
+        this.layer = null;
+
+        // color buffer format
+        this.colorFormat = this.device.defaultFramebufferAlpha ? PIXELFORMAT_R8_G8_B8_A8 : PIXELFORMAT_R8_G8_B8;
+
+        // create a depth layer, which is a default depth layer, but also a template used
+        // to patch application created depth layers to behave as one
+        if (this.device.webgl2) {
+            this.initWebGl2();
+        } else {
+            this.initWebGl1();
+        }
+    }
+
+    allocateTexture(device, name, format, isDepth, mipmaps) {
+
+        // allocate texture that will store the depth
+        const texture = new Texture(device, {
+            name,
+            format: format,
```

Comment:
Interesting - you're starting to use object shorthand....but I would use it on all properties or none. So maybe update this one to just `format`?

Reply:
and done as well, I'm warming up to it ;)